metricflow/sql/optimizer/column_pruner.py

```diff
@@ -28,171 +24,55 @@
-            pruned_from_source = node.from_source.as_select_node.accept(from_visitor)
-        else:
-            pruned_from_source = node.from_source
-        pruned_join_descriptions: List[SqlJoinDescription] = []
-        for join_description in node.join_descs:
-            right_source_as_select_node = join_description.right_source.as_select_node
-            if right_source_as_select_node:
-                right_source_visitor = SqlColumnPrunerVisitor(
-                    required_column_aliases={x.column_alias for x in right_source_as_select_node.select_columns}
-                )
-                pruned_join_descriptions.append(
-                    SqlJoinDescription(
-                        right_source=join_description.right_source.accept(right_source_visitor),
-                        right_source_alias=join_description.right_source_alias,
-                        on_condition=join_description.on_condition,
-                        join_type=join_description.join_type,
-                    )
-                )
-            else:
-                pruned_join_descriptions.append(join_description)
-
-        return SqlSelectStatementNode.create(
-            description=node.description,
-            select_columns=pruned_select_columns,
-            from_source=pruned_from_source,
-            from_source_alias=node.from_source_alias,
-            join_descs=tuple(pruned_join_descriptions),
-            group_bys=node.group_bys,
-            order_bys=node.order_bys,
-            where=node.where,
-            limit=node.limit,
-            distinct=node.distinct,
-        )
-
-    @override
-    def visit_cte_node(self, node: SqlCteNode) -> SqlQueryPlanNode:
-        raise NotImplementedError
+        self._required_alias_mapping = required_alias_mapping
 
     def visit_select_statement_node(self, node: SqlSelectStatementNode) -> SqlQueryPlanNode:  # noqa: D102
         # Remove columns that are not needed from this SELECT statement because the parent SELECT statement doesn't
         # need them. However, keep columns that are in group bys because that changes the meaning of the query.
         # Similarly, if this node is a distinct select node, keep all columns as it may return a different result set.
+        required_column_aliases = self._required_alias_mapping.get(node)
+        if required_column_aliases is None:
+            logger.error(
+                f"Did not find {node.node_id=} in the required alias mapping. Returning the non-pruned version "
+                f"as it should be valid SQL, but this is a bug and should be investigated."
+            )
+            return node
+
+        if len(required_column_aliases) == 0:
+            logger.error(
+                f"Got no required column aliases for {node}. Returning the non-pruned version as it should be valid "
+                f"SQL, but this is a bug and should be investigated."
+            )
+            return node
+
         pruned_select_columns = tuple(
```

Comment:
Sure, this could be renamed. To understand where you're coming from, you mention that `[pruned] typically is used to refer to what has been removed`. What examples were you thinking?

When I think of pruned, I think about an overgrown tree. Once I prune it, I would call it a pruned tree.

Reply:
I think that's true - the tree has been pruned. But if you're referring to the branches, I think the "pruned" branches would typically refer to the ones removed. In this case the tree is the SQL node and the columns are the branches.

I just did a quick search through the code to see how we use this word, and there are a couple places where we use the opposite meaning of prune:
https://github.com/dbt-labs/metricflow/blob/d01cbeac247c8166b26a2dbf94d727d6f1df57a4/metricflow/sql/optimizer/column_pruner.py#L37
https://github.com/dbt-labs/metricflow/blob/d01cbeac247c8166b26a2dbf94d727d6f1df57a4/tests_metricflow/sql/optimizer/test_column_pruner.py#L349-L350
https://github.com/dbt-labs/metricflow/blob/d01cbeac247c8166b26a2dbf94d727d6f1df57a4/tests_metricflow/sql/optimizer/test_column_pruner.py#L394-L395

And this is silly but I just did a quick google search for a gut check here and it does look like the pruned leaves are the ones that have been removed:
<img width="654" alt="Screenshot 2024-11-08 at 1 10 32 PM" src="https://github.com/user-attachments/assets/1b4d1e34-c1dc-4370-a173-d9cd5eca81fb">

Not totally related to this PR so this isn't blocking! But if you don't update the naming here I probably will the next time I come across it.

Reply:
In that case, I'm thinking we use different terms like "removed" and "retained" then, but will have to handle in a follow up.